File: Framework/Graphics/TextureSet.cpp
#include "stdafx.h"
#include <set>
#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>
#include "Shared/Algorithm.h"
#include "Shared/Foreach.h"
#include "Shared/StringTools.h"
#include "Shared/PathTools.h"
#include "Graphics/Texture.h"
#include "Engine/FileHandler.h"
#include "Engine/Globals.h"
#include "TextureSet.h"
#include "TextureSetReference.h"

namespace imf
{
	TextureSet::TextureSet(ImageMapPairsType& in_pairs)
	{
		std::pair<std::string, std::string> imageMapPair;
		imf_foreach (imageMapPair, in_pairs)
		{
			std::string& image(imageMapPair.first);
			std::string& map(imageMapPair.second);

			// Load texture
			boost::shared_ptr<Texture> pTexture(new Texture(image));

			// Load & parse map
			float divisor = Globals::GetHiResContentScale();
			std::vector<std::string> lines;
			FileHandler::ExtractFile(map, lines);
			CheckMsg(lines.size() > 0, "Map file non-existant or empty: " << map);

			imf_foreach (std::string& line, lines)
			{
				if (!line.empty())
				{
					std::vector<std::string> tokens;
					StringTools::Tokenize(line, "=", tokens);

					CheckMsg(tokens.size() == 2, "Invalid token count in map " << map << ", line: " << line);
					std::string idString = boost::algorithm::trim_copy(tokens[0]);
					std::string rectString = boost::algorithm::trim_copy(tokens[1]);

					std::vector<std::string> rectParts;
					StringTools::Tokenize(rectString, " ", rectParts);

					CheckMsg(rectParts.size() == 4, "Invalid rectangle token count in map " << map << ", line: " << line);

					std::string file;
					fpRect rect;

					try
					{
						file = idString;
						rect.x = ((float)boost::lexical_cast<int>(rectParts[0])) / divisor;
						rect.y = ((float)boost::lexical_cast<int>(rectParts[1])) / divisor;
						rect.w = ((float)boost::lexical_cast<int>(rectParts[2])) / divisor;
						rect.h = ((float)boost::lexical_cast<int>(rectParts[3])) / divisor;
					}
					catch (...)
					{
						CheckMsg(false, "Invalid line in map file: " << map);
					}

					CheckMsg(!imf::contains_key(textureDataSet, file), file);
					TextureData data(pTexture, rect);
					textureDataSet[file] = data;
				}
			}
		}
	}

	TextureSet::TextureSet(std::vector<std::string>& in_images)
	{
		float divisor = Globals::GetCurrentContentScale();

		// Load textures
		imf_foreach (std::string& path, in_images)
		{
			boost::shared_ptr<Texture> pTexture(new Texture(path));

			std::string fileId = PathTools::GetFilePathWithoutExtension(PathTools::GetFile(path));
			iVector2 size = pTexture->GetSize();
			iRect rect(0, 0, (float)size.x / divisor, (float)size.y / divisor);

			Check(!imf::contains_key(textureDataSet, fileId));
			TextureData data(pTexture, rect);
			textureDataSet[fileId] = data;
		}
	}

	TextureSet* TextureSet::CreateFromListFile(const std::string& in_atlasListFile)
	{
		const std::string spriteSetFolder = PathTools::GetPath(in_atlasListFile);

		std::vector<std::string> filelist;
		if (!FileHandler::ExtractFile(in_atlasListFile, filelist))
		{
			CheckMsg(false, "Texture set list file missing: " << in_atlasListFile);
		}
		
		if (!filelist.empty() && filelist[0] == "atlases:")
		{
			// Atlases
			TextureSet::ImageMapPairsType atlasPairs;

			for (uint i = 1; i < filelist.size(); ++i)
			{
				const std::string atlasPath = PathTools::Join(spriteSetFolder, filelist[i]);
				const std::string texPath = PathTools::GetFilePathWithoutExtension(atlasPath) + ".png";
				
				atlasPairs.push_back(std::make_pair(texPath, atlasPath));
			}
			
			IMF_LOG(SPRITE, INFO, "Found " << atlasPairs.size() << " atlas textures for texture set \"" << in_atlasListFile << "\".");
			
			// Load atlas
			return new TextureSet(atlasPairs);
		}
		else
		{
			// Separate textures
			imf_foreach (std::string& texPath, filelist)
			{
				texPath = PathTools::Join(spriteSetFolder, texPath);
			}
			
			IMF_LOG(SPRITE, INFO, "Found " << filelist.size() << " textures for texture set \"" << in_atlasListFile << "\".");
			
			return new TextureSet(filelist);
		}
	}

	TextureSet::~TextureSet()
	{
		textureDataSet.clear();
		// Flush textures?
	}

	bool TextureSet::GetTexture(const std::string& in_name, TextureSetReference& out_texture)
	{
		fpRect r;
		TexturesType::iterator iter = textureDataSet.find(in_name);
		if (iter != textureDataSet.end())
		{
			out_texture = TextureSetReference(in_name, name, iter->second.pTexture.get(), iter->second.area);
			return true;
		}

		return false;
	}
}

File: Framework/Graphics/TextureSet.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

using MG.Framework.Numerics;
using MG.Framework.Utility;
using MG.Framework.Converters;

namespace MG.Framework.Graphics
{
	/// <summary>
	/// A generic set of textures, primarily used for storing atlas(es).
	/// </summary>
	public class TextureSet
	{
		private struct TextureData
		{
			public RectangleF Area;
			public Texture2D Texture;

			private TextureData(Texture2D texture, RectangleF area)
			{
				Texture = texture;
				Area = area;
			}
		};

		private readonly Dictionary<string, TextureData> textureDataSet = new Dictionary<string, TextureData>();
		
		/// <summary>
		/// Create a texture set as an atlas.
		/// </summary>
		/// <param name="in_pairs">A list of image-map pairs.</param>
		public TextureSet(List<Tuple<FilePath, FilePath>> imageMapPairs)
		{
			foreach (var pair in imageMapPairs)
			{
				string image = pair.Item1;
				string map = pair.Item2;

				// Load texture
				var texture = new Texture2D(image);

				// Load & parse map
				float divisor = 1; //Globals::GetHiResContentScale();
				IEnumerable<string> lines = File.ReadLines(map);

				Debug.Assert(lines != null, "Map file non-existant or empty: " + map);

				foreach (string line in lines)
				{
					if (string.IsNullOrEmpty(line))
					{
						continue;
					}

					string[] tokens = line.Split(new[] { "=" }, StringSplitOptions.RemoveEmptyEntries);

					Debug.Assert(tokens.Length == 2, "Invalid token count in map " + map + ", line: " + line);
					string idString = tokens[0].Trim();
					string rectString = tokens[1].Trim();

					string[] rectParts = rectString.Split(new[] { " " }, StringSplitOptions.RemoveEmptyEntries);
					Debug.Assert(rectParts.Length == 4, "Invalid rectangle token count in map " + map + ", line: " + line);

					string file;
					RectangleF rect;

					file = idString;
					TypeConvert.FromString(rectParts[0], out rect.X);
					TypeConvert.FromString(rectParts[1], out rect.Y);
					TypeConvert.FromString(rectParts[2], out rect.Width);
					TypeConvert.FromString(rectParts[3], out rect.Height);
					rect.X /= divisor;
					rect.Y /= divisor;
					rect.Width /= divisor;
					rect.Height /= divisor;

					Debug.Assert(!textureDataSet.ContainsKey(file));
					textureDataSet.Add(file, new TextureData { Area = rect, Texture = texture });
				}
			}
		}

		/// <summary>
		/// Create a set from a group of images
		/// </summary>
		/// <param name="images">Images to create a set from.</param>
		public TextureSet(IEnumerable<FilePath> images)
		{
			float divisor = 1.0f; //; Globals::GetCurrentContentScale();

			// Load textures
			foreach (string path in images)
			{
				var texture = new Texture2D(path);

				string fileId = Path.GetFileNameWithoutExtension(path);
				Vector2 size = texture.Size;
				var rect = new RectangleF(0, 0, size.X / divisor, size.Y / divisor);

				Debug.Assert(!textureDataSet.ContainsKey(fileId));
				textureDataSet.Add(fileId, new TextureData { Area = rect, Texture = texture });
			}
		}

		/// <summary>
		/// Fetch a named texture from this set.
		/// </summary>
		/// <param name="textureName">Name of texture.</param>
		/// <param name="texture">Outputted texture.</param>
		/// <returns>True if the texture exists.</returns>
		public bool GetTexture(string textureName, out TextureSetReference texture)
		{
			TextureData textureData;
			if (textureDataSet.TryGetValue(textureName, out textureData))
			{
				texture = new TextureSetReference(textureName, "", textureData.Texture, textureData.Area); // TODO: Set name?
				return true;
			}

			texture = new TextureSetReference();
			return false;
		}

		public void Dispose()
		{
			foreach (var textureData in textureDataSet)
			{
				textureData.Value.Texture.Dispose();
			}
		}

		public static TextureSet CreateFromDirectory(FilePath directory)
		{
			string[] files = Directory.GetFiles(directory, "*.txt");
			
			// Atlases
			var atlasPairs = new List<Tuple<FilePath, FilePath>>();

			foreach (var file in files)
			{
				var atlasPath = new FilePath(file);
				var texPath = atlasPath.ChangeExtension("png");

				atlasPairs.Add(new Tuple<FilePath, FilePath>(texPath, atlasPath));
			}

			// Load atlas
			return new TextureSet(atlasPairs);
		}

		public static TextureSet CreateFromListFile(FilePath atlasListFile)
		{
			string spriteSetFolder = Path.GetDirectoryName(atlasListFile);

			string[] fileList = File.ReadAllLines(atlasListFile);

			if (fileList[0] == "atlases:")
			{
				// Atlases
				var atlasPairs = new List<Tuple<FilePath, FilePath>>();

				for (uint i = 1; i < fileList.Length; ++i)
				{
					FilePath atlasPath = Path.Combine(spriteSetFolder, fileList[i]);
					FilePath texPath = atlasPath.ChangeExtension("png");

					atlasPairs.Add(new Tuple<FilePath, FilePath>(texPath, atlasPath));
				}

				// Load atlas
				return new TextureSet(atlasPairs);
			}
			else
			{
				// Separate textures
				//foreach(string texPath in fileList)
				//{
				//    texPath = Path.Combine(spriteSetFolder, texPath);
				//}

				var paths = new FilePath[fileList.Length];
				for (int i = 0; i < fileList.Length; i++)
				{
					paths[i] = fileList[i];
				}

				return new TextureSet(paths);
			}
		}
	};
}